### DISCUSSION TITLE
What are the best practices for parallel inserts to a collection(s)?

### DISCUSSION THREAD

**Discussion by gaborvecsei** [milvus-io/milvus#22413] [Q&A and General discussion] [answered]
I have multiple collections and I would like to insert data into them:
- Insert data parallel to a single collection
- Insert data parallel to multiple collections

Does it work out of the box if I just create `N` threads and then start inserting data?

```python
def insert_data(data):
    collection.insert(data)
    collection.flush()
```

If this is not the case then what are the best practices?

**Comment by yhmo** [comment, marked as answer]
You don't need to call flush() every time. Each insert request will send the data to a buffer of datanode/querynode. The flush operation is to flush the buffer into storage to generate a new segment. Milvus triggers internal flush operation by time interval(by default 10 minutes). So, if you are continually performing insert requests, you don't need to call flush() manually.
If you call the flush() frequently, there will be a large number of tiny segments, and these tiny segments will bring impact to compaction machinery.

- Insert data parallel to a single collection
```
for i in range(xxx):
    collection.insert(data)

collection.flush()
```

- Insert data parallel to multiple collections
```
def insert_1():
   for i in range(xxx):
       collection_1.insert(data)

   collectin_1.flush()

def insert_2():
   for i in range(xxx):
       collection_2.insert(data)
   collectin_2.flush()

t1 = threading.Thread(target=insert_1)
t2 = threading.Thread(target=insert_2)
t1.start()
t2.start()
```

**Reply by gaborvecsei** [reply]
Thank you for clarifying the `flush()` mechanism.

To double check then this piece of code should not cause any issues, am I right?

```python
def inser_to_collection(collection, data_batch):
    ...
    collection.insert(data_batch)

collection = Collection("demo")

with future.concurrent.ThreadPoolExecutor() as executor:
    for data_batch in huge_dataset:
        executor.submit(inser_to_collection, collection, data_batch)

collection.flush()
```

**Reply by yhmo** [reply]
On the server side, Milvus handles insert requests in this way:
1. proxy node accepts insert requests via grpc channel(there is a pool in grpc to process requests asynchronously)
2. proxy node passes insert requests to pulsa/kfka (pulsar/kfka is like a WAL component for milvus)
3. once a request is successfully stored into pulsar/kafka, the insert() API will return to client, so that the client knows the insert request is accepted by server.
4. data node and query node subscribe the requests from pulsar/kafka asynchronously. The data node accumulates the data into a buffer, once a buffer is full, data node will create a new segment and flush the data into storage. The query node keeps the data in memory for rapid search.
5. the pulsar/kafka keeps the requests in their storage. Once the data node or query node crashes and restarts, it will retrieve the requests from the crash time point during restarting. But the requests are not kept in pulsar/kafka for a long time(by default, 1 day).

So, we can see:
- insert performance depends on both client-side and pulsar/kafka. On the client side, users can use multi-process or multi-thread to call insert(). On the server side, each insert request must wait the pulsar/kafka to write the request into storage.
- the pulsar/kafka will require a lot of disk space if you insert a huge dataset within one day. If the pulsar/kafka is deployed in cluster mode, you need to assign data_size*X disk space to pulsar/kafka

With our experience, for 128 dimensions vectors, the recommended way is: 30000~50000 rows for each insert call, no need to use multi-process/thread. If you want to try the max ability of pulsar/kafka, you also can use several threads/processes to insert.

By the way, for python language, there is a global lock for multi-thread, the performance is not good, multi-process is better.
